autopilot/contractor.go

```diff
@@ -221,6 +233,129 @@ func (c *contractor) performContractMaintenance(cfg api.AutopilotConfig, cs api.
 	return c.ap.bus.SetContractSet("autopilot", contractset)
 }
 
+func (c *contractor) performWalletMaintenance(cfg api.AutopilotConfig, cs api.ConsensusState) error {
+	c.logger.Info("performing wallet maintenance")
+	b := c.ap.bus
+	l := c.logger
+
+	// no contracts - nothing to do
+	numHostWanted := cfg.Contracts.Hosts
+	if numHostWanted == 0 {
+		l.Debug("wallet maintenance skipped, no contracts wanted")
+		return nil
+	}
+
+	// no allowance - nothing to do
+	if cfg.Contracts.Allowance.IsZero() {
+		l.Debug("wallet maintenance skipped, no allowance set")
+		return nil
+	}
+
+	// no money - nothing to do
+	balance, err := b.WalletBalance()
+	if err != nil {
+		return err
+	}
+	if balance.IsZero() {
+		l.Debug("wallet maintenance skipped, zero balance in wallet")
+		return nil
+	}
+
+	// fetch pending transactions
+	pending, err := c.ap.bus.WalletPending()
+	if err != nil {
+		return nil
+	}
+
+	// pending maintenance transaction - nothing to do
+	c.mu.Lock()
+	currID := c.maintenanceTxnID
+	c.mu.Unlock()
+	for _, txn := range pending {
+		if currID == txn.ID() {
+			l.Debugf("wallet maintenance skipped, pending transaction found with id %v", currID)
+			return nil
+		}
+	}
+
+	// fetch wallet outputs
+	outputs, err := b.WalletOutputs()
+	if err != nil {
+		return err
+	}
+	numOutputs := uint64(len(outputs))
+
+	// enough outputs - nothing to do
+	if numOutputs > numHostWanted {
+		l.Debugf("no wallet maintenance needed, plenty of outputs available (%v>%v)", numOutputs, numHostWanted)
+		return nil
+	}
+
+	// fetch active contracts
+	active, err := b.ActiveContracts()
+	if err != nil {
+		return err
+	}
+	numContracts := uint64(len(active))
+
+	// if enough contracts - check if we have enough outputs for refreshes
+	var missingForRefresh int
+	var missingForFormation int
+	if numContracts >= numHostWanted {
+		numOutputsForRefresh := uint64(math.Ceil(float64(numContracts) * minOutputsForRefreshesPct))
+		if numOutputs >= numOutputsForRefresh {
+			return nil
+		}
+		missingForRefresh = int(numOutputsForRefresh - numOutputs)
+	} else {
+		missingForFormation = int(numOutputs) - int(numHostWanted-numContracts)
+		missingForFormation = int(math.Max(float64(missingForFormation), 0))
+	}
```

Comment:
This seems a bit too complicated imo. Why not just keep 1 output per host that we want a contract with? I don't think that there is a lot of value in anything more fancy than that.